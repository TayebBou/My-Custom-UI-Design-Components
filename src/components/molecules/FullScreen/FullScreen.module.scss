.button {
    position: absolute;
    bottom: 0;
    right: 0;
    background-color: #00000075;
    padding: 0;
    z-index: 1;
}

.button:enabled:hover {
    background-color: #000000b0;
}

.img {
    width: 1.7em;
    display: flex;
    padding: 1em;

    @media (orientation: landscape) and (display-mode: fullscreen) {
        width: 3em;
        display: flex;
        padding: 1.5em;
    }
}

