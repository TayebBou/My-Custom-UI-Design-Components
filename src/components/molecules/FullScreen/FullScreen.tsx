import { FC } from 'react'
import Button from '../../atoms/Button/Button'
import fullScreen from '../../../assets/images/full_screen.png'
import styles from './FullScreen.module.css'
import Icon from '../../atoms/Icon/Icon'


const FullScreen: FC<{ element?: HTMLElement }> = ({ element }) => {

  const openFullScreen = () => {
    if(element) {
      if (!document.fullscreenElement) {
        element.requestFullscreen()
      } else {
        document.exitFullscreen()
      }
    } else {
      if (!document.fullscreenElement) {
        document.firstElementChild?.requestFullscreen()
      } else {
        document.exitFullscreen()
      }
    }
  }

  return (
    <Button className={styles.button} onClick={openFullScreen}>
      <Icon src={fullScreen} alt="full screen" className={styles.img} />
    </Button>
  )
}

export default FullScreen
