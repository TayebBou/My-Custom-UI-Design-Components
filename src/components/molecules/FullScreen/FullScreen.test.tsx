import { screen, render, fireEvent } from "@testing-library/react";
import FullScreen from "./FullScreen";

const givenHtmlElement = document.createElement("div");
givenHtmlElement.requestFullscreen = jest.fn();

const props = {
  element: givenHtmlElement,
  className: "FullScreen__button",
};

describe("FullScreen component", () => {
  beforeEach(() => {
    // Mock requestFullscreen method on HTMLDivElement and document.documentElement
    const requestFullscreenMock = jest.fn(() => Promise.resolve());

    // Mock requestFullscreen on document.documentElement
    Object.defineProperty(document.documentElement, "requestFullscreen", {
      value: requestFullscreenMock,
      writable: true,
    });

    // Mock exitFullscreen method explicitly on document
    document.exitFullscreen = jest.fn(() => Promise.resolve());
  });

  afterEach(() => {
    // Cleanup mocks after each test
    jest.clearAllMocks();
  });

  test("should render correctly with props", () => {
    const { asFragment } = render(<FullScreen {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render correctly without props", () => {
    const { asFragment } = render(<FullScreen />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("when user clicks fullscreen button without a given HTML element", async () => {
    // Arrange
    render(<FullScreen />);
    const fullScreenButton = await screen.findByRole("button");
    // Act: First click to enter fullscreen
    fireEvent.click(fullScreenButton);
    // Assert: Check if requestFullscreen was called on document.documentElement
    expect(document.documentElement.requestFullscreen).toHaveBeenCalled();
    // Simulate entering fullscreen
    Object.defineProperty(document, "fullscreenElement", {
      configurable: true,
      get: jest.fn(() => document.documentElement), // Now in fullscreen
    });
    // Act: Second click to exit fullscreen
    fireEvent.click(fullScreenButton);
    // Assert: Check if exitFullscreen was called
    expect(document.exitFullscreen).toHaveBeenCalled();
  });

  test("when user clicks fullscreen button with a given HTML element", async () => {
    render(<FullScreen {...props} />);
    const fullScreenButton = await screen.findByRole("button");
    // Act: First click to enter fullscreen
    fireEvent.click(fullScreenButton);
    // Assert that requestFullscreen was called on the given element
    expect(givenHtmlElement.requestFullscreen).toHaveBeenCalled();
    // Simulate entering fullscreen
    Object.defineProperty(document, "fullscreenElement", {
      configurable: true,
      get: jest.fn(() => givenHtmlElement), // Now in fullscreen
    });
    // Act: Second click to exit fullscreen
    fireEvent.click(fullScreenButton);
    // Assert that exitFullscreen was called
    expect(document.exitFullscreen).toHaveBeenCalled();
  });
});
