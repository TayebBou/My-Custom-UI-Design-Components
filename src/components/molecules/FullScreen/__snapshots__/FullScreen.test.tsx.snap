// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreen component should render correctly with props 1`] = `
<DocumentFragment>
  <button
    class="button button FullScreen__button"
    type="button"
  >
    <img
      alt="full screen"
      class="img"
      src="full_screen.png"
      style="filter: none;"
    />
  </button>
</DocumentFragment>
`;

exports[`FullScreen component should render correctly without props 1`] = `
<DocumentFragment>
  <button
    class="button button "
    type="button"
  >
    <img
      alt="full screen"
      class="img"
      src="full_screen.png"
      style="filter: none;"
    />
  </button>
</DocumentFragment>
`;
